skytemple_files/graphics/chara_wan/writer.py

```diff
@@ -0,0 +1,1412 @@
+    # Unk#6
+    out_file.write(max_blocks.to_bytes(2, 'little'))
+    # Unk#7
+    out_file.write((0).to_bytes(2, 'little'))
+    # Unk#8
+    out_file.write((0).to_bytes(2, 'little'))
+    # Unk#9
+    out_file.write((0).to_bytes(2, 'little'))
+    # Unk#10
+    out_file.write((0).to_bytes(2, 'little'))
+
+    # Image Data Info
+    ptrImageDataInfo = out_file.tell()
+    write_ptr(out_file, ptrImageDataTable, sir0_ptrs)
+    write_ptr(out_file, ptrPaletteInfo, sir0_ptrs)
+    # Unk#13
+    out_file.write((0).to_bytes(2, 'little'))
+    # Is256ColorSpr - never
+    out_file.write((0).to_bytes(2, 'little'))
+    # Unk#11
+    if len(ptrImgs) > 0:
+        out_file.write((1).to_bytes(2, 'little'))
+    else:
+        out_file.write((0).to_bytes(2, 'little'))
+    # nbImgs
+    out_file.write(len(ptrImgs).to_bytes(2, 'little'))
+
+    # WAN header
+    ptrWAN = out_file.tell()
+    write_ptr(out_file, ptrAnimInfo, sir0_ptrs)
+    write_ptr(out_file, ptrImageDataInfo, sir0_ptrs)
+    # img type - always 1 for character sprites
+    out_file.write((1).to_bytes(2, 'little'))
+    # Unk#12
+    out_file.write((0).to_bytes(2, 'little'))
+
+    # fill with AA until divisible by 16
+    padUntilDiv(out_file, b'\xAA', 16)
+
+    # sir0 pointers
+    ptrSir0 = out_file.tell()
+    for idx, offset in enumerate(sir0_ptrs):
+        offset_to_encode = offset
+        if idx > 0:
+            offset_to_encode -= sir0_ptrs[idx - 1]
+        # This tells the loop whether it needs to encode null bytes, if at least one higher byte was non-zero
+        has_higher_non_zero = False
+        # Encode every bytes of the 4 bytes integer we have to
+        for i in range(4, 0, -1):
+            currentbyte = (offset_to_encode >> (7 * (i - 1))) & 0x7F
+            # the lowest byte to encode is special
+            if i == 1:
+                # If its the last byte to append, leave the highest bit to 0 !
+                out_file.write(currentbyte.to_bytes(1, 'little'))
+            elif currentbyte != 0 or has_higher_non_zero:
+                # if any bytes but the lowest one! If not null OR if we have encoded a higher non-null byte before!
+                out_file.write((currentbyte | 0x80).to_bytes(1, 'little'))
+                has_higher_non_zero = True
+    # null terminate the sir0 header list
+    out_file.write(b'\x00')
+
+    # fill with AA until divisible by 16
+    padUntilDiv(out_file, b'\xAA', 16)
+
+    out_file.seek(4)
+    out_file.write(ptrWAN.to_bytes(4, 'little'))
+    out_file.write(ptrSir0.to_bytes(4, 'little'))
+
+
+# This will accurately load all sir0 found in m_ground, m_attack, and monster.bin with a few exceptions:
+# m_ground_0546_0xb01840.wan - Armaldo.  Metaframe Unk#0 is a nonzero
+# m_ground_0587_0xbd0a10.wan - Latios.  Unknown difference
+# monster_0433_0x3dd1c0.sir0 - Giratina Origin.  Extra zeroes after file end.
+# monster_0438_0x3ec6d0.sir0 - Shaymin Sky.  Extra zeroes after file end.
+def ImportWan(in_file):
```

Comment:
same as above, please accept bytes as input instead.
Also please use the existing Sir0 logic and if you want the `Sir0Serializable` base class for the model to make that more straight forward. Same applies to import.
It would probably also be better to move the read-in code to another module.